src/contexts/SwapContext.tsx

```diff
@@ -264,44 +263,71 @@ export const SwapContextProvider: FC<{
   const [txStatus, setTxStatus] = useState<
     | {
         txid: string;
-        status: 'loading' | 'fail' | 'success';
+        status: 'loading' | 'fail' | 'success' | 'timeout';
       }
     | undefined
   >(undefined);
 
-  const onTransaction: OnTransaction = async (txid, awaiter) => {
-    const tx = txStatus?.txid === txid ? txStatus : undefined;
-    if (!tx) {
-      setTxStatus((prev) => ({ ...prev, txid, status: 'loading' }));
-    }
-
-    const success = !((await awaiter) instanceof Error);
-
-    setTxStatus((prev) => {
-      const tx = prev?.txid === txid ? prev : undefined;
-      if (tx) {
-        tx.status = success ? 'success' : 'fail';
-      }
-      return prev ? { ...prev } : undefined;
-    });
-  };
-
   const [lastSwapResult, setLastSwapResult] = useState<ISwapContext['lastSwapResult']>(null);
   const onSubmit = useCallback(async () => {
     if (!walletPublicKey || !wallet?.adapter || !quoteResponseMeta) {
       return null;
     }
 
+    let intervalId: NodeJS.Timer | undefined;
     try {
-      const swapResult = await exchange({
-        wallet: wallet?.adapter as SignerWalletAdapter,
-        routeInfo: quoteResponseMeta,
-        onTransaction,
-        computeUnitPriceMicroLamports,
-      });
-      console.log({ swapResult });
+      const swapResult = await new Promise<SwapResult | null>(async (res, rej) => {
+        const timeout = { current: 0 };
+
+        const result = await exchange({
+          wallet: wallet?.adapter as SignerWalletAdapter,
+          routeInfo: quoteResponseMeta,
+          onTransaction: async (txid, awaiter) => {
+            if (!intervalId) {
+              intervalId = setInterval(() => {
+                if (Date.now() > timeout.current) {
+                  setTxStatus({ txid: '', status: 'timeout' });
+                  rej(new Error('Transaction timed-out'));
+                }
+              }, 1_000);
+            }
+
+            if (timeout.current === 0) {
+              timeout.current = Date.now() + 60_000;
+            }
```

Comment:
move this to above `if (!intervalId)` to prevent race condition that might happen